### DISCUSSION TITLE
Deploying Budibase in AWS Fargate

### DISCUSSION THREAD

**Discussion by kpturner** [Budibase/budibase#10374] [Help]
I want to try to get this working but I am getting lots of problems.  For simplicity I decided just to use the self contained `budibase/budibase:latest` image.

I created a ECS task template for this as below:

```
{
    "family": "self-hosted-budibase",
    "containerDefinitions": [
        {
            "name": "budibase",
            "image": "budibase/budibase:latest",
            "cpu": 0,
            "portMappings": [
                {
                    "name": "budibase-80-tcp",
                    "containerPort": 80,
                    "hostPort": 80,
                    "protocol": "tcp",
                    "appProtocol": "http"
                }
            ],
            "essential": true,
            "environment": [],
            "environmentFiles": [],
            "mountPoints": [
                {
                    "sourceVolume": "budibase_data",
                    "containerPath": "/data",
                    "readOnly": false
                }
            ],
            "volumesFrom": [],
            "logConfiguration": {
                "logDriver": "awslogs",
                "options": {
                    "awslogs-create-group": "true",
                    "awslogs-group": "/ecs/self-hosted-budibase",
                    "awslogs-region": "eu-west-1",
                    "awslogs-stream-prefix": "ecs"
                }
            }
        }
    ],
    "executionRoleArn": "blah blah/ecsTaskExecutionRole",
    "networkMode": "awsvpc",
    "volumes": [
        {
            "name": "budibase_data",
            "host": {}
        }
    ],
    "requiresCompatibilities": [
        "FARGATE"
    ],
    "cpu": "256",
    "memory": "512",
    "ephemeralStorage": {
        "sizeInGiB": 21
    },
    "runtimePlatform": {
        "cpuArchitecture": "X86_64",
        "operatingSystemFamily": "LINUX"
    }
}
```

But the container fails to start.  The logs say (repeatedly):

```
[notice] 2023-04-20T21:22:20.483831Z couchdb@127.0.0.1 <0.365.0> -------- chttpd_auth_cache changes listener died because the _users database does not exist. Create the database to silence this notice.
```
```
[error] 2023-04-20T21:22:20.484123Z couchdb@127.0.0.1 emulator -------- Error in process <0.4099.0> on node 'couchdb@127.0.0.1' with exit value:
{database_does_not_exist,[{mem3_shards,load_shards_from_db,"_users",[{file,"src/mem3_shards.erl"},{line,400}]},{mem3_shards,load_shards_from_disk,1,[{file,"src/mem3_shards.erl"},{line,375}]},{mem3_shards,load_shards_from_disk,2,[{file,"src/mem3_shards.erl"},{line,404}]},{mem3_shards,for_docid,3,[{file,"src/mem3_shards.erl"},{line,97}]},{fabric_doc_open,go,3,[{file,"src/fabric_doc_open.erl"},{line,39}]},{chttpd_auth_cache,ensure_auth_ddoc_exists,2,[{file,"src/chttpd_auth_cache.erl"},{line,198}]},{chttpd_auth_cache,listen_for_changes,1,[{file,"src/chttpd_auth_cache.erl"},{line,145}]}]}
```

So it doesn't seem to be able to create the couchDB database.  I am hoping this is something simple.

By the way. Fargate networkMode awsvpc does not allow port mapping. Is there an envvar that can be used with this docker image to make it listen on something other than port 80?

**Comment by kpturner** [comment]
What does a label of "Out of scope" mean?

I am more interested in what sort of situation would make the container fail in this way. I can sort out the AWS Task definition accordingly.

**Reply by melohagan** [reply]
Hey @kpturner The tag is simply indicating that the content of the question is beyond the customer success team's knowledge (e.g. me!)

We're predominantly here to help with Budibase platform specific questions, e.g. the app builder, app management, etc.

The issue is still open though if anyone else is able to help out!

**Reply by kpturner** [reply]
There is a thread regarding AWS somewhere that seems to have gone quiet - should I add it to that?  I just need to find out what causes BB to fail to create its own couchdb database.

**Reply by shogunpurple** [reply]
Hi @kpturner - this may be of use to you for deploying budibase on fargate:

https://github.com/neilkuan/cdk-budibase

We use AWS but do not run the budibase cloud stack ourselves in fargate. A couple of things:

- I think you can do port mapping in your setup, can you not just change the `hostPort` to something else?
- Your container `cpu` value is `0` in your config, you should set it to 256 like the task and restart

If this doesn't work:
There's a script that runs as part of our couchDB setup that should create the `_users` database etc for you. There's a chance that this is not running for some reason and provisioning the couchDB database inside the single image. 

When you start the image, is there absolutely nothing else in the logs? Just the couch log?

**Comment by kpturner** [comment]
Thanks.  As  I mentioned in the OP, you can’t map the port in Fargate - it’s forbidden by the mandatory `awsvpc` networkMode (the only mode you can have with Fargate). If you change hostPort to something else the task creation just fails immediately.  The `cpu` and `memory` definitions are at the task level (at the bottom). The `cpu` definition for the container is optional for Fargate so I do not think it’s related to that. I think it’s something to do with ephemeral storage.  The logs at the beginning need more analysis - I’ll try again but not wait so long for the creation event to fail.

**Comment by imvikysharma** [comment]
It looks like the container is failing to start because it cannot find the "_users" database in CouchDB. One possible solution is to create the "_users" database manually in CouchDB before starting the container. You can do this by logging into the CouchDB web interface, clicking on the "Create Database" button, and entering "_users" as the database name.

**Reply by kpturner** [reply]
I don't think that is possible in a scenario like this - the container is not available in the ECS task until the Fargate service has been started and deemed healthy.  And that won't be the case until the container is listening on its port. So there is not a CouchDB instance to log in to. 

 I think it is down to this error in the log:
 ```

23/04/2023, 19:32:26 | chown: cannot access '/data/couch': No such file or directory
-- | --

```
which tells me the mount point is causing the problem.
```
            "mountPoints": [
                {
                    "sourceVolume": "budibase_data",
                    "containerPath": "/data",
                    "readOnly": false
                }
            ],
```
```
"networkMode": "awsvpc",
"volumes": [
        {
            "name": "budibase_data",
            "host": {}
        }
    ],
```
ECS Fargate provides 20GB of ephemeral storage to the task (which in and of itself would be an issue for multiple instance - but I can cross that bridge later).  So I really need to figure out the correct volume definitions so that the budibase container has a volume to create the database in.

**Comment by kpturner** [comment]
I have been comparing the cloudwatch logs from the Fargate instance to those I see when running the BB container locally.  The local instance exhibits all the same errors as the Fargate instance does.  It too reports on several occasions that the _users database does not exist - but I guess there is some background process spawned to create it which eventually works and things continue as normal.  However, if such a process exists, it is not reporting anything in the Cloudwatch logs as to why it cannot create the database.  The _users database just never appears, so the logs I see about it being missing just continue on forever. I do at least see `couch_replicator_clustering : cluster stable` which is somewhat positive. 

@shogunpurple is there any way to see what the script does and to work out if it is even starting?  Are there ny obvious logs that it can (be made to) produce?

**Comment by andygrillo** [comment]
did you resolve this @kpturner  ? I am about to start running this on ECS also.

**Reply by kpturner** [reply]
No I am afraid I gave up.  Ran out of time.

**Reply by andygrillo** [reply]
ok, thats a shame, thanks

**Comment by hunsche** [comment]
Hi everyone,

I'd like to share a solution that worked perfectly for me when dealing with this issue. I followed the instructions from the [neilkuan/cdk-budibase](https://github.com/neilkuan/cdk-budibase) repository and made a few small modifications.

Specifically, I set the user as '0' in the code, as follows:

```javascript
createAcl: {
  ownerGid: '0',
  ownerUid: '0',
  permissions: '0755',
},
posixUser: {
  uid: '0',
  gid: '0',
},
```

With these settings, everything worked beautifully.

Additionally, @andygrillo I'd like to mention that it's shameful to complain without taking any steps to solve the problem. If everyone contributes their solutions and experiences, we can continually improve the community and the project.

I hope this helps :rocket:
